### PR TITLE
Sleepers version 2.2.0

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
- use latest Dockerfile best practices
- standard sleeper is again below 150MB
- sleeper-mpi/-gpu use latest Cuda 12.3.1
- adds dream input to generate a file of any size in bytes
- adds dream output as file of the dream input size for testing
- simplify generation of synchronized sleeper/-mpi/-gpu with 1 single metadata file
- upgraded Github workflow file
- version bumped to 2.2.0
- uses latest uv version in Github CI

### details
- sleeper uses:
  - python 3.11-slim image
- sleeper-mpi/-gpu uses
  - nvidia/cuda-12.3.1-base-ubuntu-22.04 image

## Checklist

- [ ] I think the code is well written
- [ ] I updated the table of contents (make toc)
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] **Runs in the swarm**
- [ ] If you design a new module, add your user to .github/CODEOWNERS

### CODE REVIEW

File: services/sleeper/validation/output/outputs.json [hunk 1]
```diff
@@ -1,4 +1,5 @@
 {
     "output_1": "single_number.txt",
-    "output_2": 2
-}
+    "output_2": 2,
+    "output_3": "dream.txt"
```
Comment (pcrespov) [inline services/sleeper/validation/output/outputs.json:4]: haha `nightmare.txt` at times :-)

File: services/sleeper/docker/custom/Dockerfile [hunk 2]
```diff
@@ -33,10 +72,21 @@ ENV PIP_EXTRA_INDEX_URL=https://www.piwheels.org/simple
...
+    pip install uv~=0.1
 
-# Instlling python
-RUN pip3 install --upgrade pip wheel setuptools
+# NOTE: python virtualenv is used here such that installed
+# packages may be moved to production image easily by copying the venv
+RUN uv venv "${VIRTUAL_ENV}"
+
+RUN --mount=type=cache,mode=0755,target=/root/.cache/pip \
+    uv pip install --upgrade  \
```
Comment (pcrespov) [inline services/sleeper/docker/custom/Dockerfile:86]: haha uv man!

File: services/sleeper/metadata/metadata.yml [hunk 3]
```diff
@@ -68,3 +79,10 @@ outputs:
       title: Random sleep interval
       type: integer
       x_unit: second
+  output_3:
+    displayOrder: 3
+    label: Dream output
+    description: Contains some random data representing a dream
```
Comment (pcrespov) [inline services/sleeper/metadata/metadata.yml:85]: very good description :-D

In order to be more accurate, perahps you can read [How the brain constructs dreams](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7279884/#:~:text=Deep%20inside%20the%20temporal%20lobe,to%20remember%2C%20imagine%20and%20dream.&text=Our%20most%20vivid%20dreams%20are,a%20richly%20detailed%20hallucinatory%20experience.) and provide the right inputs ;-)

**Review by pcrespov — APPROVED** [review]
thx for this big cleanup!

File: services/sleeper/docker-compose-meta.yml [hunk 4]
```diff
@@ -5,23 +5,27 @@ services:
...
+          "anderegg@itis.swiss", "affiliation": "IT''IS Foundation"}]}'
         io.simcore.contact: '{"contact": "maiz@itis.swiss"}'
         io.simcore.description: '{"description": "A service which awaits for time
           to pass, two times."}'
         io.simcore.inputs: '{"inputs": {"input_1": {"displayOrder": 1, "label": "File
           with int number", "description": "Pick a file containing only one integer",
           "type": "data:text/plain", "fileToKeyMap": {"single_number.txt": "input_1"}},
           "input_2": {"displayOrder": 2, "label": "Sleep interval", "description":
-          "Choose an amount of time to sleep in range [0-65]", "defaultValue": 2,
+          "Choose an amount of time to sleep in range [0-900]", "defaultValue": 2,
```
Comment (matusdrobuliak66) [inline services/sleeper/docker-compose-meta.yml:17]: Q: Why are giving an upper limit? I guess we might test at some point even running the job for 3 hours?

**Review by matusdrobuliak66 — APPROVED** [review]
Thanks for the upgrade! :) please consider putting even higher upper limit (so we can run it let say for 3 hours?) thanks!

File: services/sleeper/docker-compose-meta.yml (hunk 4 above)
Comment (sanderegg) [reply services/sleeper/docker-compose-meta.yml:17]: not sure who started with that upper limit... I can make it go to a day or 2

Comment (sanderegg) [reply services/sleeper/docker-compose-meta.yml:17]: actually removing the max

File: services/sleeper/validation/input/inputs.json [hunk 5]
```diff
@@ -2,5 +2,6 @@
     "input_1": "single_number.txt",
     "input_2": 2,
     "input_3": false,
-    "input_4": 4
-}
+    "input_4": 4,
+    "input_5": 1024
+}
```
Comment (bisgaard-itis) [inline services/sleeper/validation/input/inputs.json:7]: Here again

**Review by bisgaard-itis — APPROVED** [review]
Very nice! Thanks a lot

File: services/sleeper/metadata/metadata.yml (hunk 3 above)
Comment (elisabettai) [reply services/sleeper/metadata/metadata.yml:85]: We also have neuroscientists in the Team than can help. 😇

**Review by elisabettai — APPROVED** [review]
Thanks for the deep (sleep) refactoring.

@wvangeit might like it a bit more, since it the code does something (i.e. producing a dream) instead of just waiting for time to pass. 🙃

File: services/sleeper/service.cli/run [hunk 6]
```diff
@@ -18,6 +18,8 @@ INPUT_3=$(< "$json_input" jq '.input_3')
 export INPUT_3
 INPUT_4=$(< "$json_input" jq '.input_4')
 export INPUT_4
+INPUT_5=$(< "$json_input" jq '.input_5')
+export INPUT_5
 
 exec execute.sh
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/sleeper/service.cli/run:25]: it indicates an empty line with spaces it seems... will fix.

File: services/sleeper/validation/input/inputs.json (hunk 5 above)
Comment (sanderegg) [reply services/sleeper/validation/input/inputs.json:7]: well here my linter removes the empty line

**Review by GitHK — APPROVED** [review]

**Pull request merged**